packages/remix-dev/compiler/plugins/cssFilePlugin.ts

```diff
@@ -0,0 +1,103 @@
+import * as path from "path";
+import * as fse from "fs-extra";
+import esbuild from "esbuild";
+
+import { BuildMode } from "../../build";
+import type { BuildConfig } from "../../compiler";
+
+const isExtendedLengthPath = /^\\\\\?\\/;
+
+const normalizePathSlashes = (path: string) =>
+  isExtendedLengthPath.test(path) ? path : path.replace(/\\/g, "/");
+
+/**
+ * This plugin loads css files with the "css" loader (bundles and moves assets to assets directory)
+ * and exports the url of the css file as its default export.
+ */
+export function cssFilePlugin(
+  buildConfig: Pick<Partial<BuildConfig>, "mode">
+): esbuild.Plugin {
+  return {
+    name: "css-file",
+
+    async setup(build) {
+      let buildOps = build.initialOptions;
+
+      build.onLoad({ filter: /\.css$/ }, async (args) => {
+        let { outfile, outdir, assetNames } = buildOps;
+        let { metafile, outputFiles, warnings, errors } = await esbuild.build({
+          ...buildOps,
+          minify: buildConfig.mode === BuildMode.Production,
+          minifySyntax: true,
+          metafile: true,
+          write: false,
+          sourcemap: false,
+          incremental: false,
+          splitting: false,
+          stdin: undefined,
+          outfile: undefined,
+          outdir: outfile ? path.dirname(outfile) : outdir,
+          entryNames: assetNames,
+          entryPoints: [args.path],
+          loader: {
+            ...buildOps.loader,
+            ".css": "css",
+          },
+          // this plugin treats absolute paths in 'url()' css rules as external to prevent breaking changes
+          plugins: [
+            {
+              name: "resolve-absolute",
+              async setup(build) {
+                build.onResolve({ filter: /.*/ }, async (args) => {
+                  let { kind, path: resolvePath } = args;
+                  if (kind === "url-token" && path.isAbsolute(resolvePath)) {
+                    return {
+                      path: resolvePath,
+                      external: true,
+                    };
+                  }
+                });
+              },
+            },
+          ],
+        });
+
+        if (errors) {
```

Comment:
I think we need to change to `errors && errors.length !== 0` here. I've patched remix with this patch, and give it try. But failed with `import xx from "katex/dist/katex.min.css"`, finding that `errors = []`.

Reply:
@KingSora  @xhebox - would this CSS plugin allow us to import css files like this at the top of JS files, similar to CSS modules?

```ts
import "./styles.css"
// or
import styles from "./style.css";
```